Add support for pollutant setters.

abhiramm7 opened the pull request.

abhiramm7 commented:
@cbuahin I am running into an error with reg testing. Is this one of the errors that is fixed in the 5.1.14? 
<img width="971" alt="Screen Shot 2021-02-20 at 4 28 27 PM" src="https://user-images.githubusercontent.com/14915561/108608981-ad9e7b80-7398-11eb-94b7-3e12bcaf5d72.png">

michaeltryby commented:
This error occurred because the requirements were ignored and the latest version of SWMM-toolkit got installed which is incompatible with the nrtest-SWMM package.

abhiramm7 commented:
> This error occurred because the requirements were ignored and the latest version of SWMM-toolkit got installed which is incompatible with the nrtest-SWMM package.

@michaeltryby got it. I’ll just wait for the 5.1.14 merge. I’ll  fix the tolerance error  in Ubuntu.

cbuahin commented:
> > This error occurred because the requirements were ignored and the latest version of SWMM-toolkit got installed which is incompatible with the nrtest-SWMM package.
> 
> @michaeltryby got it. I’ll just wait for the 5.1.14 merge. I’ll fix the tolerance error in Ubuntu.

@abhiramm7 , you could try reverting to the older version of the swmm toolkit.

michaeltryby commented on src/solver/qualrout.c:
```diff
@@ -330,8 +364,22 @@ void findLinkQual(int i, double tStep)
+
+            Link[i].newQual[p] = Link[i].extQual[p];
+            Link[i].extPollutFlag[p] = 0;
+	}
```
Conceptually, having difficulty understanding the necessity of associating an external pollutant with a link. How would a pollutant enter a link?

michaeltryby replied:
I get it now. This is for modeling an eroding channel. Or a channel that has accumulated a large amount of sediment that is getting resuspended during storm events.

michaeltryby commented on src/solver/toolkit.c:
```diff
@@ -1874,6 +1874,8 @@ int DLLEXPORT swmm_getNodeResult(int index, SM_NodeResult type, double *result)
             case SM_LATINFLOW:
                 *result = Node[index].newLatFlow * UCF(FLOW); break;
+	    case SM_HRT:
+		*result = Node[index].hrt; break;
```
Is SM_HRT an input parameter or a result of a calculation?

abhiramm7 replied:
HRT is computed by the SWMM and this exposes the computed HRT to users.

michaeltryby commented on tests/solver/CMakeLists.txt:
```diff
@@ -33,7 +33,7 @@ set_target_properties(test_lid
     test_canonical.cpp
     test_gage.cpp
-    test_pollut.cpp
+    test_pollutants.cpp
```
Why is rename necessary?

abhiramm7 replied:
Every thing else was fully spelt, so for consistency I updated the spelling from `test_pollut.cpp`

michaeltryby reviewed (commented):
Hey @abhiramm7 , Lots of questions. Hopefully they make sense and we can have a back and forth discussion getting into the details. Thanks!

dickinsonre commented on src/solver/include/toolkit_enums.h:
```diff
@@ -132,12 +132,15 @@ typedef enum {
     SM_NODEHEAD       = 6,  /**< Node Head */
-    SM_LATINFLOW      = 7   /**< Lateral Inflow Rate */
+    SM_LATINFLOW      = 7,  /**< Lateral Inflow Rate */
+    SM_HRT	      = 8   /**< HRT */
```
HRT is more obvious  and matches the treatment equations

void updateHRT(int j, double v, double q, double tStep)
//
//  Input:   j = node index
//           v = storage volume (ft3)
//           q = inflow rate (cfs)
//           tStep = time step (sec)
//  Output:  none
//  Purpose: updates hydraulic residence time (i.e., water age) at a 
//           storage node.

abhiramm7 replied:
Most of the pollutant treatment literature uses the term HRT, so we stuck with it.

abhiramm7 commented on src/solver/objects.h:
```diff
@@ -485,6 +485,7 @@ typedef struct
    char*         ID;              // node ID
    int           type;            // node type code
    int           subIndex;        // index of node's sub-category
+   int*          extPollutFlag;   // external pollutant flag
```
External pollutant is a property of the node. `extQual` variable is the container for storing pollutant value that user wants to set in the node and `extPollutantFlag` controls when the value in `extQual` is set in the node. If it is 0, pollutant in the node is computed based on the pollutants in the model. When the flag is 1, pollutant in the node is set to the value in `extQual`

jennwuu replied:
@abhiramm7 What is reason for not modifying the existing pollutant concentration within SWMM model?

abhiramm7 replied:
@jennwuu If I understand your question right, are you asking why we are not directly replacing the pollutant variable? If so, it has to do with how pollutants are handled in the swmm. When swmm computes pollutant transformations, it stores the calculated values in the `cOut` variable, uses it to do mass balance, and sets it as newQual. If we were to set newQual from an external function call directly, we would have to reimplement mass balance. By updating `cOut` we can build on the existing mass balance functionality in swmm. Furthermore, if we don't have a container variable and have to set these pollutants at each step directly, we would have to interrupt the treatment function, make an external function call to python, read the variable, set it, and repeat this every node and pollutant. By setting the pollutants in a container before swmm step is called, we can limit the number of external function calls we have to make, thus reducing the computational overhead.

Digital-Water replied:
hello Abhiramm, have you applied the SWMM model to simple sellement lagoon? I assume the problem is working out settlement rates for slow moving water and the ground infitration?

abhiramm7 replied:
Hi @Digital-Water, the idea behind this PR was to provide us a way implement the slow settling behavior in basins. We have a few implementations in https://github.com/kLabUM/StormReactor/blob/master/StormReactor/waterQuality.py that you might find interesting. It is not exactly a settlement lagoon, but I think we can build on these.

abhiramm7 commented on src/solver/objects.h:
```diff
@@ -512,6 +514,9 @@ typedef struct
    double*       newQual;         // current quality state
+   double*	 extQual;	  // external quality state
+   double*	 inQual;          // inflow quality state
+   double*	 reactorQual;     // concentration in the mixed reactor   
```
`extQual` is the container for the pollutant that the user wants to set in the node. CSTR and PFR pollutant models need inflow and reactor concentration for mass balance, hence we track both `inQual` and `reactorQual`.

jennwuu replied:
Thinking through this, can you use oldQual variable to get inQual?

abhiramm7 replied:
Good question @jennwuu, but oldQual would account for dilution in the node or link, so we had to create a new variable for inQual

bemason commented:
Are there still outstanding questions that need to be answered so that we can merge these revisions?

jennwuu commented:
> Are there still outstanding questions that need to be answered so that we can merge these revisions?

I will try to provide a PR review by June.

Thank you for your patience @bemason. 

In the meantime, can you update the PR so the tests pass? Thanks!

abhiramm7 commented:
Hi @katmratliff, if you have time can you review this PR. It adds support for updating pollutants in the nodes and the links during the simulation.

jennwuu reviewed (commented):
@abhiramm7 Since new variables are proposed to represent external pollutants, can you provide schematic of how external pollutant travel within network? 

Can you also provide explanation for adding external pollutant versus adding getter/setter for swmm pollutant?

abhiramm7 commented:
@jennwuu thanks for the comments! regarding documentation, we have a package https://github.com/kLabUM/StormReactor that documents how to use this features. The plan is for us to submit a PR to pyswmm once these additions are approved with getter and setters for pollutant. Hopefully this addresses your comment. If I misunderstood something, please let me know.

abhiramm7 commented:
Regarding on how this works, I should have included the previous discussion in closed PR. This PR should provide more context https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/pull/302#issue-483016814. Please let me know if you have any additional questions. Thanks!

jennwuu commented on src/solver/objects.h:
```diff
@@ -634,6 +639,7 @@ typedef struct
    char*         ID;              // link ID
    int           type;            // link type code
    int           subIndex;        // index of link's sub-category
+   int*          extPollutFlag;   // external pollutant flag
```
why is this a pointer?

jennwuu replied:
isn't it just 1 or 0 for state?

abhiramm7 replied:
You are right it is 0 or 1. But given that it is set by an external function by reference, we had to define it as a pointer.

jennwuu commented on src/solver/qualrout.c:
```diff
@@ -128,11 +130,15 @@ void qualrout_execute(double tStep)
     for ( i = 0; i < Nobjects[LINK]; i++ ) findLinkQual(i, tStep);
+
+    // --- default to swmm treatment
+    if (ExtPollutFlag == 1){ExtPollutFlag = 0;}
```
Can you explain the purpose of this? Where is ExtPollutFlag defined? What is it turned off if it is was turned on previously.

abhiramm7 replied:
For sure, ExtPollutFlag is created when a pollutant object is created for the node or link. It controls what pollutant to set in the node. If 0, which is set by default, it uses the pollutant routing in swmm and when 1 it sets newqual to the value in extqual.

As the end of each step, we reset flag to 0 so that if user does not provide a pollutant then it go back to swmm routing than breaking down.

jennwuu commented on src/solver/qualrout.c:
```diff
@@ -330,8 +337,22 @@ void findLinkQual(int i, double tStep)
+            lossExtQual = lossExtQual * v1/ tStep;
+            massbal_addReactedMass(p, lossExtQual);
+
+            Link[i].newQual[p] = Link[i].extQual[p];
```
@abhiramm7 I might be misunderstanding something. I was under the impression that external pollutant was separate from pollutant variables in swmm. But Line 353 is showing that newQual is being updated by extQual. 

Back to the original request, I think it would be really helpful to have docs for the new feature in this PR.

abhiramm7 replied:
For sure, Brooke and I are working on doc sheet. Will add it as soon as we are done.

michaeltryby commented:
@abhiramm7 I've got it building and running on Ubuntu 18.04 running in WSL. Still getting the link errors running tests in Actions. Confusing as all get out.

michaeltryby commented:
@abhiramm7 tests are build and running on all platforms. The tests themselves are failing on Windows, and Ubuntu. Passing on MacOS. Please evaluate changes I made to testing criteria. Thanks!

michaeltryby commented:
@abhiramm7 what’s going on with the quality mass balance errors?

abhiramm7 commented:
> Hey @abhiramm7! Hope you had a good trip. When I was working on the unit tests I noticed some tests with checks on quality mass balance errors that seemed quite high. Are you experiencing mass balance issues with these modifications to the code?

Hi @michaeltryby :). It was a nice break. Yeah, mass balance for pollutant links are quite high. I have been digging into it yesterday. One reason might be that I am drastically adding pollutants in links where there was nothing. I have reduced the pollutants that I add (+10 and -9 mg/l) and they seem be under control (< 0.10%).

abhiramm7 commented:
@jennwuu we've added documentation in the wiki (https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/wiki/Pollutant-Module:-get-and-set-pollutants-during-simulation), and test are passing. Please do let me know if you want us to add anything else. Thanks :). CC: @bemason

jennwuu commented:
@abhiramm7 Great documentation! That is exactly what I am looking for. Give me 2-3 days to review and I will provide more feedback. Thanks!

michaeltryby commented:
Or maybe you need to update the mass balance when you add or remove mass from the system?

abhiramm7 commented:
> Or maybe you need to update the mass balance when you add or remove mass from the system?

@michaeltryby Yeah, nodes were fairly straight forward as they were already built to handle pollutant removal.. For links we used the same logic SWMM used for first order mixing, https://github.com/abhiramm7/Stormwater-Management-Model/blob/0d8db401dad3b225fd300bc4c743c4228233e387/src/solver/qualrout.c#L349

bemason commented:
Hi @jennwuu just checking in to see if you need any additional info from us! Hoping to keep this moving along! :)

jennwuu commented:
@bemason I am still reviewing the documentation. I will provide feedback as soon as I can

jennwuu reviewed (changes requested):
Looking through the documentation, can you update the documentation for the following points?

Documentation Format Feedback
- can you adjust format section: current pollutant process? it looks incorrect on my screen
- fix all hyperlinks within documentation
- fix typo Simulation 5 ii) 
- Python snippet under use case would be more useful if we provide the code in text format instead of screenshot
- fix format of section: use case section (the bullet points are in one paragraph)
- move use case at bottom of wiki doc 
- under variables section, can you indicate which variables are NEW to SWMM engine?
- under setter section, introduce the function and the parameters similar to getters section
- update the setter/getter names so that they match the code function names 
- include a section to document how mass balance is maintained with extQual variables 

Questions
- Answered via share screen review
- Why is reactorQual an integer? 
- Add one test to set concentration to zero for entire simulation and confirm mass balance

abhiramm7 commented:
> @abhiramm7 Let me know when you finalize the comments. I can do a final review.
> Thanks!

@jennwuu We have updated the documentation and added a new test to address your comments. Please do let me know if you have any additional comments. Thanks :)

jennwuu reviewed (approved):
The doc looks good to me. Thank you for addressing all my concerns!

michaeltryby commented:
Congratulations @abhiramm7! Great to see this merged!

abhiramm7 commented:
@michaeltryby and @jennwuu thanks for all your help along the way 😁
